### PR TITLE
implement rebase

### PR DESCRIPTION
**Pull request by phritz**

### CODE REVIEW

File: src/embed/types.rs [hunk 1]
```diff
@@ -44,8 +52,12 @@ pub struct CloseTransactionRequest {
 #[derive(SerJson)]
 pub struct CloseTransactionResponse {}
 
-#[derive(DeJson)]
-pub struct GetRootRequest {}
+#[derive(Debug, DeJson, SerJson)]
+pub struct GetRootRequest {
+    #[nserde(rename = "headName")]
+    #[nserde(skip_serializing_if = "Option::is_none")]
+    pub head_name: Option<String>,
```
Comment (aboodman) [inline src/embed/types.rs:59]: What brought on this change? replicache-client doesn't support this field:

https://github.com/rocicorp/replicache-client/blob/master/repm/types.go#L14

File: src/db/commit.rs [hunk 2]
```diff
@@ -102,16 +102,19 @@ impl Commit {
         self.commit().value_hash().unwrap()
     }
 
-    pub fn next_mutation_id(&self) -> u64 {
+    pub fn mutation_id(&self) -> u64 {
         let meta = self.commit().meta().unwrap();
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/db/commit.rs:106]: done in https://github.com/rocicorp/repc/pull/193

File: src/embed/connection.rs [hunk 3]
```diff
@@ -198,27 +199,24 @@ async fn do_open<'a, 'b>(
         Some(mutator_name) => {
             let OpenTransactionRequest {
                 name: _,
-                rebase_opts,
                 args: mutator_args,
+                rebase_opts,
             } = req;
+            let mutator_args = mutator_args.ok_or(ArgsRequired)?;
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/embed/connection.rs:205]: i dont think args are required for read transactions but are for write transactions. if they have a mutator name they should also have args.

File: src/embed/connection.rs [hunk 4]
```diff
@@ -240,6 +238,52 @@ async fn do_open<'a, 'b>(
...
+) -> Result<(), OpenTransactionError> {
+    use OpenTransactionError::*;
+
+    // TODO should we also enforce here that the basis is equal to the sync head?
+    let (_, original, _) = db::read_commit(db::Whence::Hash(opts.original_hash.clone()), &dag_read)
+        .await
+        .map_err(NoSuchOriginal)?;
+    match original.meta().typed() {
+        db::MetaTyped::Local(lm) => {
+            if lm.mutator_name() != mutator_name {
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/embed/connection.rs:255]: think we talked about this but to refresh: yes it's a sanity check on the bindings. the mutation returned to be replayed should be the mutation replayed.

File: src/embed/connection.rs [hunk 5]
```diff
@@ -253,10 +297,18 @@ async fn do_commit<'a, 'b>(
         Transaction::Write(w) => Ok(w),
         Transaction::Read(_) => Err(TransactionIsReadOnly),
     }?;
-    txn.commit(db::DEFAULT_HEAD_NAME, "local-create-date")
+    let head_name = match txn.is_rebase() {
+        false => db::DEFAULT_HEAD_NAME,
+        true => sync::SYNC_HEAD_NAME,
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/embed/connection.rs:302]: back then who knows but now as you recall we error out the earlier sync the next time it makes a call

File: src/embed/types.rs (hunk 1 above)
Comment (phritz) [reply src/embed/types.rs:59]: huh, guessing i wanted to use it in a dispatch test but then something happened and i removed the use of it. someone later added it to the rpc. can remove it if you want tho seems useful for poking at things.

**Pull request merged**